### PR TITLE
Implement pasting external content with plain and styled text

### PR DESCRIPTION
**Pull request by mkevins**
## Description

This PR aims to implement paste handling [(targets 1 and 2)](https://github.com/wordpress-mobile/gutenberg-mobile/issues/238#issuecomment-454779457) for the Mobile Gutenberg project. It is part of the work to address [#238](https://github.com/wordpress-mobile/gutenberg-mobile/issues/238). The goal is to respond to a paste event from ReactNative code and transform the data for use with Gutenberg's existing paste handling implementations.

## How has this been tested?

This PR is a work in progress, and so far testing has been manual ([some manual tests done for Android and known issues are described here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/553#issue-250970841)). When the implementation is more complete / stable, mock events will be created, as well as tests for the implementations.

A local test environment has been set up using the `./bin/setup-local-env.sh` script.

## DOM implementation on mobile

In order to implement these changes, it was necessary to modify several areas of the codebase. This mostly involved changes to `*.native.js` (Mobile Gutenberg only) files, but some shared code as well. In particular, a few extra safety checks have been added for situations where properties or methods are null or undefined in the ReactNative environment due to an incomplete implementation of DOM classes.

The codepaths utilizing some DOM methods can be hard to follow at times. As an example, when the `pasteHandler` function calls `deepFilterHTML`, it uses the `specialCommentConverter` filter. The `raw-handling/special-comment-converter.js` file has a [dependency on `@wordpress/dom`](https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/raw-handling/special-comment-converter.js#L4), and makes use of the `remove` and `replace` functions from there.

There is a comment in the `rawHandler` function describing the use of the `specialCommentConverter` filter there: `// Needed to create more and nextpage blocks.` As `pasteHandler` uses the same filter and calls through to the same function (`deepFilterHTML`) I believe this comment's description applies to the filter's use in `pasteHandler` as well.

Deeper into this codepath, where `deepFilterNodeList` makes use of the filters (the forEach loop), is where the need arises for the `contains` method (which had to be patched into `jsdom-jscore`). In the forEach loop, `item` is actually the filter function.

## Types of changes
* For more flexibility in code-sharing, I refactored `pasteHandler` from `raw-handling/index.js` into its own file.
* I added safety checks for `jsdom` methods/properties.
* I refactored `splitContent` for use with `onEnter` and `onPaste`, and implemented `createRecord` for converting native content to the RichText format used in several other methods.
* I implemented `onPaste` for plain and styled text.
* I mirrored the implementation of `onReplace` from `edit.js` in `edit.native.js`, and pasted image blocks are now functional in the demo app.

## Notes

Originally, I had planned on approaching this task in stages as described in the original issue. However, I realized upon implementing plain text paste, that styled text was implemented with very little additional effort.

**Edit:** Similarly, images are now working with the above approach, after implementing `onReplace`.

## Related PRs
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/553

### CODE REVIEW

**Comment by diegoreymendez** [comment]
@mkevins - I'll be testing these changes through `gutenberg-mobile`, and will let you know if there's anything.

One suggestion I wanted to make for future PRs, is to avoid moving large blocks of code between files, unless the PR is specifically about that.

The reason behind this suggestion is that during the review the diff becomes a mix of code-changes and code-movements that becomes hard to differentiate, and as such, hard to validate.

**Comment by mkevins** [comment]
> @mkevins - I'll be testing these changes through `gutenberg-mobile`, and will let you know if there's anything.

Thank you for testing it out!

> One suggestion I wanted to make for future PRs, is to avoid moving large blocks of code between files, unless the PR is specifically about that.
> 
> The reason behind this suggestion is that during the review the diff becomes a mix of code-changes and code-movements that becomes hard to differentiate, and as such, hard to validate.

Thanks for the suggestion. I was torn on this, as it wasn't clear which would be more noisy, splitting on PRs with multiple related PRs across projects, or reducing the number of PRs and clarifying the changes in the commit message. I refactored (the big code movements) in the first commit and isolated those changes by commit. I'll keep this in mind in future PRs, and try to keep them smaller.

If you have any questions, or want clarification on any part of the code, I'll be happy to help.

**Comment by mkevins** [comment]
I added a commit to refactor `rawHandler` back into `raw-handling/index.js` (rather than its own file), as it is not required on the native side for this implementation. Hopefully, this will make reviewing the PR easier.

**Comment by hypest** [comment]
👋 @iseulde , can you make a pass on this PR? It touches some files used from the web too so, just want to make sure we don't break anything there. Thanks!

File: packages/block-library/src/paragraph/edit.native.js [hunk 1]
```diff
@@ -71,6 +72,20 @@ class ParagraphEdit extends Component {
 		}
 	}
 
+	onReplace( blocks ) {
```
Comment (hypest) [inline packages/block-library/src/paragraph/edit.native.js]: This seems like fully copy pasted from the web side module so, let's extract it into its own module and resuse it both on the web and native. What do you think?

Comment (mkevins) [reply packages/block-library/src/paragraph/edit.native.js]: I had a similar thought, my initial hesitation was that the method calls a reference to the parent component method of the same name, and the shadowed (inner scope) reference is to the parent version, so I wasn't sure if that would be easy to follow in a separate file. The more I think on it, your suggestion makes sense. I'll try it and push a commit.

File: packages/blocks/src/api/raw-handling/phrasing-content-reducer.js [hunk 2]
```diff
@@ -4,7 +4,9 @@
 import { wrap, replaceTag } from '@wordpress/dom';
 
 export default function( node, doc ) {
-	if ( node.nodeName === 'SPAN' ) {
+	// In jsdom-jscore, 'node.style' can be null.
```
Comment (hypest) [inline packages/blocks/src/api/raw-handling/phrasing-content-reducer.js]: Kinda a nit but, perhaps it's not too clear what `jsdom-jscore` means in this context. Of course, jsdom-jscore it's the DOM parser we're using in the native mobile (React Native) version of Gutenberg, but I think I'd suggest add that bit to the comment to, to help out web-side friends read the code more easily.

So, my suggestion is to refer with `In native mobile (jsdom-jscore), ...` instead of the current. What do you think? Same for a couple other similar references in this same PR.

Comment (hypest) [reply packages/blocks/src/api/raw-handling/phrasing-content-reducer.js]: On a less of a nit thing, I wonder about the (emphasis mine on the "can"):

```
'node.style' *can* be null.
```

Does that mean that sometimes node.style is undefined and some other it's not?

In the cases where it's undefined, is it OK to just ignore it like we're doing currently in the PR? I wonder if being undefined denotes an unrecoverable error and the we should let the user know about it, in some way.

Should we at least log an error/warning?

File: packages/blocks/src/api/raw-handling/paste-handler.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
+/**
```
Comment (iseulde) [inline packages/blocks/src/api/raw-handling/paste-handler.js]: Why is moving to a separate file needed?

File: packages/editor/src/components/rich-text/index.native.js [hunk 4]
```diff
@@ -246,10 +254,123 @@ export class RichText extends Component {
...
+			start: selectionStart,
+			end: selectionEnd,
+			...create( {
+				html: innerContent,
+				range: null,
+				multilineTag: false,
+				removeNode: null,
+				unwrapNode: null,
+				removeAttribute: null,
+				filterString: null,
```
Comment (iseulde) [inline packages/editor/src/components/rich-text/index.native.js]: The above four parameters have been removed.

File: packages/blocks/src/api/raw-handling/paste-handler.js (hunk 3 above)
Comment (hypest) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: 👋 @iseulde , during a Slack chat with @mkevins , I suggested that in order to limit scope, we should extract only the method we currently need on mobile, which is the `pasteHandler()` method. That's then exported by the `index.native.js` module.

Comment (hypest) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: It's then used in `rich-text/index.native.js`.

Comment (iseulde) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: Ok. I see any the moving of web code but... nothing else has changed. The extra checks look ok.

File: packages/blocks/src/api/raw-handling/phrasing-content-reducer.js (hunk 2 above)
Comment (mkevins) [reply packages/blocks/src/api/raw-handling/phrasing-content-reducer.js]: Yes, good suggestion! I'll edit those, and keep that in mind for future comments.

File: packages/editor/src/components/rich-text/index.native.js (hunk 4 above)
Comment (mkevins) [reply packages/editor/src/components/rich-text/index.native.js]: Thanks for letting me know. Will remove.

File: packages/blocks/src/api/raw-handling/paste-handler.js (hunk 3 above)
Comment (hypest) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: I understand you completed your review pass @iseulde . Thanks for the quick response! 🙇‍♂️

Comment (mkevins) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: An `index.native.js` file already existed for `raw-handling`, and I was not able to import the vanilla `pasteHandler` from `index.js`.

File: packages/dom/package.json [hunk 5]
```diff
@@ -19,6 +19,7 @@
 	},
 	"main": "build/index.js",
 	"module": "build-module/index.js",
+	"react-native": "src/index",
```
Comment (hypest) [inline packages/dom/package.json]: Since this effectively means we're starting to use the `@wordpress/dom` package, I wonder if you have some insight on what are we using it for? Like, I understand is for _pasting_, but I wonder what parts of the process are actually using it.

It's not easy to follow the code to see where exactly it's used so, having some general idea is still useful if you have it.

File: packages/editor/src/components/rich-text/index.native.js [hunk 6]
```diff
@@ -92,35 +98,32 @@ export class RichText extends Component {
 	 * handler.
 	 *
 	 */
-	splitContent( htmlText, start, end ) {
+	splitContent( currentRecord, blocks = [], isPasted = false ) {
 		const { onSplit } = this.props;
 
 		if ( ! onSplit ) {
```
Comment (hypest) [inline packages/editor/src/components/rich-text/index.native.js]: Not strictly related to this PR but, it seems that the `onSplit` prop is already marked as deprecated on the web side so, since we're touching the `splitContent` method, what do you think about mirroring this (https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/rich-text/index.js#L84-L93) here too?

I'd suggest a follow up PR with that since it's not a blocker for this PR and adding more diff to this is not desirable :)

File: packages/editor/src/components/rich-text/index.native.js [hunk 7]
```diff
@@ -92,35 +98,32 @@ export class RichText extends Component {
 	 * handler.
 	 *
 	 */
-	splitContent( htmlText, start, end ) {
+	splitContent( currentRecord, blocks = [], isPasted = false ) {
```
Comment (hypest) [inline packages/editor/src/components/rich-text/index.native.js]: > isPasted = false

I understand that we just need this flag here but, may I suggest copying what the web side is providing the corresponding method with? The idea is to stay as close as possible to the web code, wishing we can refactor and re-use the same code eventually.

I this case, let's pass a `context` object with a `paste` (boolean) field. WDYT?

**Review by hypest — APPROVED** [review]
👋 @mkevins , great work here, works really nicely!

I left a few comments which I think could use some minor work. Good thing is, I don't see any blockers for this PR in any of those! I'd suggest we tackle them all in follow up PR(s) and go ahead and merge this one. Let me know if that's OK with you too.

If OK with you, then please open a ticket with a checklist of all the known issues and additional work needed so we can more easily work against them.

File: packages/dom/package.json (hunk 5 above)
Comment (mkevins) [reply packages/dom/package.json]: When the `pasteHandler` function calls `deepFilterHTML`, it uses the `specialCommentConverter` filter. The `raw-handling/special-comment-converter.js` file has a [dependency on `@wordpress/dom`](https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/raw-handling/special-comment-converter.js#L4), and makes use of the `remove` and `replace` functions from there.

There is a comment in the `rawHandler` function describing the use of the `specialCommentConverter` filter there: `// Needed to create more and nextpage blocks.` As `pasteHandler` uses the same filter and calls through to the same function (`deepFilterHTML`) I believe this comment's description applies to the filter's use in `pasteHandler` as well.

By the way, deeper into this codepath, where `deepFilterNodeList` makes use of the filters (the forEach loop) is where the need arises for the `contains` method (which had to be patched into jsdom-jscore). In the forEach loop, `item` is actually the filter function.

FYI: If you had run the bundler previously (`yarn start`), and subsequently removed the line in question `"react-native": "src/index",` from `packages/dom/package.json`, you will not run into an error until you run `yarn start:reset`.

File: packages/block-library/src/paragraph/edit.native.js (hunk 1 above)
Comment (mkevins) [reply packages/block-library/src/paragraph/edit.native.js]: I have created [this commit](https://github.com/mkevins/gutenberg/commit/1fcd88c134f9094bdb227b5b273dd833cc5ac011) on a new branch which refactors this. After creating it, I'm conflicted, as I wonder about the pattern introduced. Originally, the function is defined as a prototype method, but is bound to `this` in the constructor (a common pattern, but effectively making it a new function, occupying its own place in memory). With this commit, it will also need be bound to `this` in the constructor, so I don't think there will be any impact on memory. However, it differs from the pattern used elsewhere, where we refactor stand-alone functions that were already being exported from an index file (whereas here we are refactoring an internal prototype method).

Having `onReplace` as a copy-pasted prototype method in both the vanilla and native versions (i.e. leaving it as it is) is not DRY, so modifications / maintenance requires more work. On the other hand, I'm uncertain whether changes to the vanilla implementation would always be safe for the native side, so maybe there is some benefit to the manual effort required to mirror such modifications / changes. I don't have a strong opinion one way or the other, but I just wanted to document my thoughts as we may encounter similar situations in other areas of the codebases. WDYT?

A few notes: 
- I had to use an alias for parent onReplace function in the standalone file to satisfy linting rules, but I think this actually clarifies the code.
- I had to remove onReplace from some of the destructured assignments, and use the more verbose `this.props.onReplace` for the same linting reasons.

File: packages/blocks/src/api/raw-handling/phrasing-content-reducer.js (hunk 2 above)
Comment (mkevins) [reply packages/blocks/src/api/raw-handling/phrasing-content-reducer.js]: > Does that mean that sometimes node.style is undefined and some other it's not?

With the change of wording in the comment to: `In native mobile (jsdom-jscore), ...`, it's probably more accurate to say that `node.style` *is* undefined.

> Should we at least log an error/warning?

In the codepath where this is used, `node` is actually an `HTMLElement`, as the condition is satisfied that the element's `nodeName` is `SPAN`. HTMLElement nodes are supposed to have a `style` property which returns a read-only [`CSSStyleDeclaration`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration) object, which can be used to retrieve the inline styles applied to the element via the style attribute. This is then used to wrap the node in appropriate tag(s) (`strong`, `em`, etc.). In the case where the DOM implementation does not support the `style` property on a node, I believe it is safe to ignore, but that we may lose style information that was associated with the content via inline styles on a span's attribute.

The property seems to have had wide support for a long time, but I wonder if there are any browsers that don't support this property, or that leave it undefined or null when no style attribute was present on the element. My reason for wording the comment somewhat "vaguely" is that I tend to assume that there could be such a case in the wild, and thus the safety check could be useful even in the browser implementation. I checked the [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style#Browser_compatibility) for it on MDN, and it lists a few question marks, but that just means unconfirmed AFAIK.

**TL;DR:**

The check will not lead to an unrecoverable error, but will gracefully degrade the feature that normalizes some inline styles on span tags. I don’t have a strong opinion on whether that warrants logging a warning; I suppose it depends on the verbosity desired (e.g. web: “please use a better browser :)”, mobile: "mobile currently lacks support for this feature").

File: packages/blocks/src/api/raw-handling/paste-handler.js (hunk 3 above)
Comment (mkevins) [reply packages/blocks/src/api/raw-handling/paste-handler.js]: Thank you @iseulde for having a look at my proposed changes, and for your feedback.

File: packages/block-library/src/paragraph/edit.native.js (hunk 1 above)
Comment (hypest) [reply packages/block-library/src/paragraph/edit.native.js]: I see, thanks for drilling down on this @mkevins! It seems that extracting this function ends up complicating the code, perhaps because the function is rather trivial in the first place.

I had a look at the reason it was introduced ([see this commit](https://github.com/WordPress/gutenberg/commit/a9142441669ab422338190311ee6fdc000ff5c2d)) and I think the function has little chances to get revised.

So, I think it should be OK to just have it copy-pasted from the web-side module, like you currently have it in the PR. Is that OK with you too?

File: packages/blocks/src/api/raw-handling/phrasing-content-reducer.js (hunk 2 above)
Comment (hypest) [reply packages/blocks/src/api/raw-handling/phrasing-content-reducer.js]: > mobile: "mobile currently lacks support for this feature"

I think that's a good case to actually log since, these things are useful while offering support to users.

While at it, since this is a known limitation, let's add a "Known Limitations" section to https://github.com/wordpress-mobile/gutenberg-mobile/pull/553 and mention this there. WDYT?

File: packages/dom/package.json (hunk 5 above)
Comment (hypest) [reply packages/dom/package.json]: Ah, nice! This is solid info so, let's add it to the PR description, cool?

> you will not run into an error until you run `yarn start:reset`

Right, makes sense. Metro caches the packages resolution anyway, so we need to reset the cache to pick up changes in the `package.json`s.

File: packages/editor/src/components/rich-text/index.native.js (hunk 4 above)
Comment (mkevins) [reply packages/editor/src/components/rich-text/index.native.js]: I've removed them. Thanks again!

File: packages/block-library/src/paragraph/edit.native.js (hunk 1 above)
Comment (mkevins) [reply packages/block-library/src/paragraph/edit.native.js]: Yes, I'll leave it as it is.

**Review by hypest — APPROVED** [review]
Looks good to me! Nice work @mkevins, thanks!

I'll wait for the restarted Travis job to get green and I will merge this PR.

When you get the chance, please open a ticket in the Gutenberg-mobile repo with a checklist of all the known issues and additional work needed so we can more easily work against them.

**Comment by mkevins** [comment]
> When you get the chance, please open a ticket in the Gutenberg-mobile repo with a checklist of all the known issues and additional work needed so we can more easily work against them.

I've added this issue to track the related work:
https://github.com/wordpress-mobile/gutenberg-mobile/issues/624
